src/wallet/rpcwallet.cpp

```diff
@@ -609,12 +615,15 @@ UniValue ListaddressesForPurpose(const std::string strPurpose)
     UniValue ret(UniValue::VARR);
     {
         LOCK(pwalletMain->cs_wallet);
-        for (const auto& addr : pwalletMain->mapAddressBook) {
-            if (addr.second.purpose != strPurpose) continue;
+        CAddressBookIterator it = pwalletMain->NewAddressBookIterator();
+        while (it.HasNext()) {
+            auto addrBook = it.GetValue();
+            if (addrBook.purpose != strPurpose) continue;
```

Comment:
Needs to be very careful with `continue` when changing a `for` loop to a `while`.
If the purpose doesn't match, we never reach `it.Next()`, therefore we have an infinite loop.

Reply:
upps yeah, great catch.